### PR TITLE
Add classes for forming acquisition functions

### PR DESCRIPTION
**Pull request by Thomas-Christie**
## Type of changes

- [ ] Bug fix
- [X] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [X] I've formatted the new code by running `poetry run pre-commit run --all-files --show-diff-on-failure` before committing.
- [X] I've added tests for new code.
- [X] I've added docstrings for the new code.

## Description

Added an `AbstractAcquisitionFunctionBuilder` abstract base class for forming acquisition functions, and added `ThompsonSampling` as a concrete example. Also added an `AbstractAcquisitionMaximizer`abstract base class for maximising acquisition functions, as well as a concrete implementation in the form of the `ContinuousAcquisitionMaximizer` which uses `L-BFGS-B` from the `jaxopt` library to optimise continuous acquisition functions over a bounded domain.

Also renamed the `bayes_opt` module to `decision_making` to reflect the fact that we aim for this module to support general sequential decision making, and not solely Bayes opt.

Unit tests have been added for all newly introduced components.

Issue Number: N/A

### CODE REVIEW

File: gpjax/decision_making/acquisition_functions/base.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    @abstractmethod
+    def build_acquisition_function(
+        self,
+        posteriors: Mapping[str, AbstractPosterior],
+        datasets: Optional[Mapping[str, Dataset]],
+        key: KeyArray,
+    ) -> AcquisitionFunction:
+        """
+        Build an `AcquisitionFunction` from a set of posteriors and datasets.
+        """
```
Comment (thomaspinder) [inline gpjax/decision_making/acquisition_functions/base.py:56]: Can we detail the role of each argument in this method please? It's quite important to do so in the abstract class so that users wanting to subclass it know what each argument should do.

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+from gpjax.decision_making.utils import OBJECTIVE
+from gpjax.gps import ConjugatePosterior
+from gpjax.typing import KeyArray
+
+
+@dataclass
+class ThompsonSampling(AbstractAcquisitionFunctionBuilder):
+    """
+    Form an acquisition function by drawing an approximate sample from the posterior,
+    using decoupled sampling as introduced in [Wilson et. al. (2020)](https://arxiv.org/abs/2002.09309).
```
Comment (thomaspinder) [inline gpjax/decision_making/acquisition_functions/thompson_sampling.py:32]: Would recommend adding this to the `citation.py`

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        key: KeyArray,
+    ) -> AcquisitionFunction:
+        if OBJECTIVE not in posteriors.keys():
+            raise ValueError("Objective posterior not found in posteriors")
+        if OBJECTIVE not in datasets.keys():
+            raise ValueError("Objective dataset not found in datasets")
+        if key is None:
+            raise ValueError(
+                "Key must be provided to draw an approximate sample from the posterior"
+            )
```
Comment (thomaspinder) [inline gpjax/decision_making/acquisition_functions/thompson_sampling.py:50]: Could this go into the abstract method?

File: gpjax/decision_making/acquisition_maximizer.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        best_initial_sample_point = _get_discrete_maximizer(
+            initial_sample_points, acquisition_function
+        )
+
+        # Jaxopt minimizer requires a function which returns a scalar. It calls the
+        # acquisition function with one point at a time, so the acquisition function
+        # returns an array of shape [1, 1], so  we index to return a scalar. Note that
+        # we also return the negative of the acquisition function - this is because
+        # acquisition functions should be *maximimized* but the Jaxopt minimizer
+        # minimizes functions.
```
Comment (thomaspinder) [inline gpjax/decision_making/acquisition_maximizer.py:120]: Can you add this to a docstring here please?

File: gpjax/decision_making/acquisition_maximizer.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        def scalar_acquisition_fn(x: Float[Array, "1 D"]) -> ScalarFloat:
```
Comment (thomaspinder) [inline gpjax/decision_making/acquisition_maximizer.py:121]: Maybe add an underscore before this function to make it explicit that it's private to the scope of the method

File: gpjax/decision_making/utils.py
Comment (thomaspinder) [inline gpjax/decision_making/utils.py:1]: If this is _just_ for typing, then I would rename it so.

File: tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+from gpjax.likelihoods import (
+    Gaussian,
+    Poisson,
+)
+from gpjax.mean_functions import Zero
+
+config.update("jax_enable_x64", True)
+
+
+def generate_dummy_1d_dataset() -> Dataset:
```
Comment (thomaspinder) [inline tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py:45]: For this generator and all below, could you add a seed value argument e.g., `seed_value` that then feeds in the key i.e., `jr.PRNGKey(seed_value)`? All subsequent unit tests could then be validated for more than one seed/dataset.

**Review by thomaspinder — CHANGES_REQUESTED** [review]
This is looking great @Thomas-Christie! Left a few comments.

File: gpjax/decision_making/__init__.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
-from gpjax.bayes_opt import search_space
+from gpjax.decision_making import search_space
```
Comment (henrymoss) [inline gpjax/decision_making/__init__.py:15]: nice!

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+
+@dataclass
+class ThompsonSampling(AbstractAcquisitionFunctionBuilder):
+    """
+    Form an acquisition function by drawing an approximate sample from the posterior,
+    using decoupled sampling as introduced in [Wilson et. al. (2020)](https://arxiv.org/abs/2002.09309).
+    """
+
+    num_rff_features: int = 500
```
Comment (henrymoss) [inline gpjax/decision_making/acquisition_functions/thompson_sampling.py:35]: Thats a v large default. I've had a lot of success with just 10. But perhaps thats a bit too low. Any thoughts @thomaspinder ?

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py (hunk 2 above)
Comment (henrymoss) [inline gpjax/decision_making/acquisition_functions/thompson_sampling.py:32]: Maybe worth stressing that it is the negative of the sample that is returned?

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py (hunk 3 above)
Comment (henrymoss) [reply gpjax/decision_making/acquisition_functions/thompson_sampling.py:50]: keep the key one in TS though. many other acqs are not stochastic

File: gpjax/decision_making/acquisition_maximizer.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+@dataclass
+class ContinuousAcquisitionMaximizer(AbstractAcquisitionMaximizer):
+    """The `ContinuousAcquisitionMaximizer` class is used to maximize acquisition
+    functions over the continuous domain with L-BFGS-B. First we sample the acquisition
+    function at `num_initial_samples` points from the search space, and then we run
+    L-BFGS-B from the best of these initial points.
+    """
+
+    num_initial_samples: int
```
Comment (henrymoss) [inline gpjax/decision_making/acquisition_maximizer.py:96]: it would be nice to have a num_restarts here.

For now this could just be done in a for loop, i.e. do this whole optimisation (random init + opt) a few times and return best

File: tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+        + 10 * jr.normal(jr.PRNGKey(42), shape=X0.shape)
+    )
+    D = Dataset(X, y)
+    return D
+
+
+def generate_dummy_1d_test_X(num_test_points: int) -> Float[Array, "N 1"]:
+    test_X = jnp.linspace(0.0, 10.0, num_test_points).reshape(-1, 1)
+    return test_X
+
```
Comment (henrymoss) [inline tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py:70]: this and the bits below are gonna be useful for testing all acquisition functions. Maybe pop in their own file. Perhaps @thomaspinder has a better idea? We could do something like the automatic way the kernels are tested

File: tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        posteriors=posteriors, datasets=datasets, key=key_two
+    )
+    test_X = jnp.linspace(0.0, 10.0, 100).reshape(-1, 1)
+    ts_acquisition_function_one_values = ts_acquisition_function_one(test_X)
+    ts_acquisition_function_two_values = ts_acquisition_function_two(test_X)
+    assert isinstance(ts_acquisition_function_one, Callable)
+    assert isinstance(ts_acquisition_function_two, Callable)
+    assert not (
+        ts_acquisition_function_one_values == ts_acquisition_function_two_values
+    ).all()
```
Comment (henrymoss) [inline tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py:218]: You could do a monte-carlo test here?

i.e. eval a  bunch of TS at a set of points X then check that the avg over the TS matches with the posterior mean over X?

**Review by henrymoss — APPROVED** [review]
LGTM. A couple minor Qs

**Comment by henrymoss** [comment]
I guess you could update the notebook to include this functionality?

File: gpjax/decision_making/utils.py
Comment (Thomas-Christie) [reply gpjax/decision_making/utils.py:1]: I'm intending on adding a little extra functionality to this file in subsequent commits, so will leave with the current title for now.

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py (hunk 8 above)
Comment (Thomas-Christie) [reply gpjax/decision_making/acquisition_functions/thompson_sampling.py:35]: I think the default in Trieste is 1000?

File: tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py (hunk 6 above)
Comment (Thomas-Christie) [reply tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py:45]: Good idea, was using `linspace` to generate the test points, but will also make this random to improve quality of testing.

File: tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py (hunk 11 above)
Comment (Thomas-Christie) [reply tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py:218]: Decided against doing this in the end as the test would take too long to run, and there is already a monte-carlo test for testing out the quality of approximate samples.

File: tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py (hunk 10 above)
Comment (Thomas-Christie) [reply tests/test_decision_making/test_acquisition_functions/test_thompson_sampling.py]: Have refactored so that the standard functions I introduced for testing out the acquisition maximizer (e.g. `Forrester` etc.) can also be used here.

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py (hunk 8 above)
Comment (thomaspinder) [reply gpjax/decision_making/acquisition_functions/thompson_sampling.py:38]: Yeah 500 seems very high. Usually use 50-100 whenever I use RFFs.

**Review by thomaspinder — COMMENTED** [review]
Nice work @Thomas-Christie. Once we've resolved the failing tests and the RFF comment, I'd be happy to approve.

File: gpjax/decision_making/acquisition_functions/thompson_sampling.py (hunk 8 above)
Comment (Thomas-Christie) [reply gpjax/decision_making/acquisition_functions/thompson_sampling.py]: Ok, have changed to 100. This is also consistent with the default provided in the `sample_approx` method.

**Review by thomaspinder — APPROVED** [review]
